File: 216.combination-sum-iii.js
/*
 * @lc app=leetcode id=216 lang=javascript
 *
 * [216] Combination Sum III
 */

// @lc code=start
/**
 * @param {number} k
 * @param {number} n
 * @return {number[][]}
 */
var combinationSum3 = function (k, n) {
  const helper = (cur, idx, target) => {
    if (target < 0) {
      return;
    }
    if (cur.length === k) {
      if (target === 0) {
        res.push(cur.slice());
      }
      return;
    }
    for (let i = idx; i < 10; ++i) {
      cur.push(i);
      helper(cur, i + 1, target - i);
      cur.pop();
    }
  };
  const res = [];
  helper([], 1, n);
  return res;
};
// @lc code=end

File: 216.combination-sum-iii.py
#
# @lc app=leetcode id=216 lang=python3
#
# [216] Combination Sum III
#

# Time:  O()
# Space: O()

from typing import List


# @lc code=start
class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        def helper(cur: list, idx, target):
            if target < 0:
                return
            if len(cur) == k:
                if target == 0:
                    res.append(cur[:])
                return
            for i in range(idx, 10):
                cur.append(i)
                helper(cur, i + 1, target - i)
                cur.pop()

        res = []
        helper([], 1, n)
        return res


# @lc code=end
